packages/cli/src/models/network/NetworkController.ts

```diff
@@ -313,11 +319,17 @@ export default class NetworkController {
 
   // Contract model || SolidityLib model
   private _getAllSolidityLibNames(contractNames: string[]): string[] {
+    let newLibNames = this._getSolidityLibNames(contractNames);
+    // deploy children before parents
+    // we want to deploy each contract only once
+    return uniq([...(newLibNames.length ? this._getAllSolidityLibNames(newLibNames) : []), ...newLibNames]);
+  }
```

Comment:
Very succinct!

One thing though: there may be a deadlock here.  It's possible to compile code like so:

```solidity
pragma solidity ^0.5.10;

import "./TestLib2.sol";

library TestLib1 {
  function test() public pure returns (string memory) {
    return TestLib2.test();
  }
}
```

```solidity
pragma solidity ^0.5.10;

import "./TestLib1.sol";

library TestLib2 {
  function test() public pure returns (string memory) {
    return TestLib1.test();
  }
}
```

Here TestLib1 depends on TestLib2 which depends on TestLib1 etc.  Compiling it is possible, even if deploying is not.  Further, there may be arbitrary depths of cycles: imagine if TestLib1 -> TestLib2 -> TestLib3 -> TestLib1 -> TestLib2 ... and so on.

One answer to this could be that the function _getAllSolidityLibNames has a second parameter which is all current parent dependencies.  If you could remove all the parent dependencies then you can determine what lib hasn't been searched already.

```typescript
private _getAllSolidityLibNames(contractNames: string[], allDependencies: string[] = []): string[] {
  let newLibNames = this._getSolidityLibNames(contractNames);

  let newLibs = difference(newLibNames, allDependencies)
  let repeatLibs = intersection(newLibNames, allDependencies)
  let oldLibs = difference(allDependencies, newLibNames)

  // reorder so that dependencies are pulled up
  let reorderedOldLibs = repeatLibs.concat(oldLibs)

  // create a new list of all libs
  let allLibs = newLibs.concat(reorderedOldLibs)

  return uniq([...(newLibs.length ? this._getAllSolidityLibNames(newLibs, allLibs) : []), ...reorderedOldLibs]);
}
```

I think that should do it.

Reply:
Damn. I completely missed the fact that you can compile code which wouldn't deploy. Thanks for pointing out that!